### PR TITLE
Encode username and token URL parameters

### LINKED ISSUE #114: URL Encode Usernames In URLs
We need to URL encode usernames and probably project names in URLs. This seems like something Lapis should handle. 

The specific bug is in URLs generated for password resets and activations that are sent -- this should be handled first, but I'd love if there was a generic way.

### PR DESCRIPTION
**Pull request by namoscato**
Fixes #114 

This leverages [LuaSocket's `url.escape()`](http://w3.impa.br/~diego/software/luasocket/url.html#escape) and simplifies URL generation via [Lapis's `url_for()`](http://leafo.net/lapis/reference/actions.html#request-object-methods/url_for).

### CODE REVIEW

**Comment by cycomachead** [comment]
I totally missed this; thanks so much! I'll test and pull soon. :)

File: validation.lua [hunk 1]
```diff
@@ -118,5 +124,12 @@ create_token = function (self, purpose, username, email)
         email,
         mail_subjects[purpose] .. username,
         mail_bodies[purpose],
-        self:build_url('/users/' .. username .. '/' .. purpose .. '/' .. token_value))
+        self:build_url(self:url_for(
+            purpose,
+            {
+                username = url.escape(username),
```
Comment (cycomachead) [inline validation.lua]: This works well, though it seems to encode characters that are "URL safe", like `-`. Do you know if there's a way around that? If not, this is still a clear improvement, but just trying to consider some future uses.

File: api.lua [hunk 2]
```diff
@@ -191,7 +191,7 @@ app:match('resendverification', '/users/:username/resendverification', respond_t
     end)
 }))
 
-app:match('resetpassword', '/users/:username/password_reset(/:token)', respond_to({
+app:match('password_reset', '/users/:username/password_reset(/:token)', respond_to({
```
Comment (cycomachead) [inline api.lua]: Can we add a comment noting that this must match the DB types for tokens.

File: validation.lua (hunk 1 above)
Comment (namoscato) [reply validation.lua]: Yeah, good call. [`url.build_path()`](http://w3.impa.br/~diego/software/luasocket/url.html#build_path) is what we want.

**Comment by cycomachead** [comment]
Awesome thanks!

File: api.lua [hunk 3]
```diff
@@ -191,9 +191,11 @@ app:match('resendverification', '/users/:username/resendverification', respond_t
...
     -- Methods:     GET, POST
     -- Description: Handles password reset requests.
+    --              The route name should match the database token purpose.
+    -- @see validation.create_token
```
Comment (cycomachead) [inline api.lua]: BTW, this was a smart idea!

File: validation.lua [hunk 4]
```diff
@@ -118,5 +124,12 @@ create_token = function (self, purpose, username, email)
         email,
         mail_subjects[purpose] .. username,
         mail_bodies[purpose],
-        self:build_url('/users/' .. username .. '/' .. purpose .. '/' .. token_value))
+        self:build_url(self:url_for(
+            purpose,
+            {
+                username = url.build_path({username}),
```
Comment (cycomachead) [inline validation.lua]: For some reason this function doesn't encode `=` and `&`, which I think makes this a problem for query strings, but maybe there's some happy medium?

I guess I need to do more research.

**Pull request merged**